cmd/snap-update-ns/utils.go

```diff
@@ -108,6 +115,64 @@ func secureMkdirAll(name string, perm os.FileMode, uid, gid int) error {
 	return nil
 }
 
+func designWritableMimic(dir string) ([]*Change, error) {
+	// We need a place for "safe keeping" of what is present in the original
+	// directory as we are about to attach a tmpfs there, which will hide
+	// everything inside.
+	logger.Debugf("create-writable-mimic %q", dir)
+	safeKeepingDir := filepath.Join("/tmp/.snap/", dir)
+
+	var changes []*Change
+
+	// Bind mount the original directory elsewhere for safe-keeping.
+	changes = append(changes, &Change{
+		Action: Mount, Entry: mount.Entry{
+			// XXX: should we rbind here?
+			Name: dir, Dir: safeKeepingDir, Options: []string{"bind"}},
+	})
+	// Mount tmpfs over the original directory, hiding its contents.
+	changes = append(changes, &Change{
+		Action: Mount, Entry: mount.Entry{Name: "none", Dir: dir, Type: "tmpfs"},
+	})
+	// Iterate over the items in the original directory (nothing is mounted _yet_).
+	entries, err := ioutilReadDir(dir)
```

Comment:
I'm not sure about using `ioutil.ReadDir` here. I think a better approach would be to `f, err := os.Open(dir)` and then range over `f.Readdir(<some-sane-limit))`. IIRC `iotuil.ReadDir()` does the same, but it will go and load `FileInfo` for all entries in directory. This may (but does not need to) be a problem if there are lots of files or dirs inside.

Reply:
I'm happy to tweak this though I don't think it will be a real problem. This is also nicer for mocking (one vs two + stateful helper).